### PR TITLE
Console

### PR DESCRIPTION
**Pull request by ricardoquesada**
A helper class that let's the user control the game remotely.
Currently it supports these commands: "fps on", "fps off", "scene graph", "help", "exit".

But it is easy to add new commands without modifying the cocos2d code.

It also adds a helper function in Scheduler that lets you run an `std::function<void()>` in the cocos2d thread.

Use cases:
- It can use for realtime debugging (useful for the new renderer)
- for QA automation

### CODE REVIEW

**Comment by ricardoquesada** [comment]
Test on Mac, iOS, Linux and Android.

Haven't tested it on Windows. And most probably it won't work on Windows since Windows has some strange API calls for sockets. 
One option is to disable this feature for Windows until windows support is added.

File: cocos/2d/CCScheduler.cpp [hunk 1]
```diff
@@ -950,8 +947,43 @@ void Scheduler::update(float dt)
...
+        }
+    }
+
+    //
+    // Functions allocated from another thread
+    //
+
+    // Testing size is faster than locking / unlocking.
+    // And almost never there will be functions scheduled to be called.
+    if( _functionsToPerform.size() > 0 ) {
```
Comment (dumganhar) [inline cocos/2d/CCScheduler.cpp]: I think `!empty()` is faster than `size() > 0`.

**Comment by dumganhar** [comment]
I will test it on Windows. Thanks :).

**Comment by dumganhar** [comment]
Could we use this `console` for `JSDebugger` ?

File: cocos/2d/CCScheduler.cpp (hunk 1 above)
Comment (ricardoquesada) [inline cocos/2d/CCScheduler.cpp]: fixed. thanks.

**Comment by ricardoquesada** [comment]
Yes, actually one very useful Use Case is the JS Debugger.

We can replace part of the logic of JS Debugger with the new Console.
We might need to add a few functions in the `Console` in order to support the JS Debugger properly, but I think it is worth doing it.

File: cocos/base/CCConsole.cpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        sched->performFunctionInCocosThread( std::bind(&Director::setDisplayStats, dir, true));
+    } },
+    { "fps off", [](int anFd) {
+        Director *dir = Director::getInstance();
+        Scheduler *sched = dir->getScheduler();
+        sched->performFunctionInCocosThread( std::bind(&Director::setDisplayStats, dir, false));
+    } },
+    { "scene graph", std::bind(&Console::commandSceneGraph, this, std::placeholders::_1) },
+    { "exit", std::bind(&Console::commandExit, this, std::placeholders::_1) },
+    { "help", std::bind(&Console::commandHelp, this, std::placeholders::_1) }, }
```
Comment (dumganhar) [inline cocos/base/CCConsole.cpp]: VS2012 doesn't support `initializer list`. So this usage will cause compilation errors on windows.

**Comment by dumganhar** [comment]
Great.
Then, lots of `socket connection, thread switch` codes could be removed from `ScriptingCore`.

**Comment by dumganhar** [comment]
Hi @ricardoquesada 
Could you let me know how to test this `console`, so I could make sure it also works on windows?

**Comment by dumganhar** [comment]
Got it. I used `telnet 127.0.0.1 5678` and logon successfully. :)

**Comment by ricardoquesada** [comment]
Thanks for merging it.

Regarding the JS Debugger, one possible way to implement it is by subclassing `Console` and adding all the needed functionality there.

`Console` was designed in a way that it is possible to have multiple instances of it (each one is going to run on its own thread listening to a different TCP port).

**Pull request merged**